### PR TITLE
Trigger and set: fix default flow assignment for n=0 tasks.

### LINKED ISSUE #6361: Errant flow merge when triggering
### Description

```
R1 = """
    foo
    bar
"""
```

If `foo` fails in flow 1 while `bar` is in flow 2, triggering `foo` without providing a flow in the options causes it to run in flows 1 & 2.

This is because the default `--flow all` is causing it to pick all active flow numbers in the entire task pool instead of all active flow numbers for the task in question.

https://github.com/cylc/cylc-flow/blob/293c3ccba1877adc1061064b410ea54b62beb4ad/cylc/flow/task_pool.py#L2077-L2078

https://github.com/cylc/cylc-flow/blob/293c3ccba1877adc1061064b410ea54b62beb4ad/cylc/flow/task_pool.py#L2025-L2037

Looking at the code, this bug probably affects `cylc set` too

### PR DESCRIPTION
**Pull request by hjoliver**
Close #6361

**FIX**
  - by default (i.e., `--flow` option not used), existing (n=0) tasks should keep their existing flow numbers

**Tidy**
  - remove redundant option checking, now done during command validation
  - clarify command and schema help

**NEW**
  - explicit use of `--flow` will merge the given flow numbers into existing (n=0) tasks
     (this seems sensible to me, and I can't think of a reason not to do it)

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
(The integration test fails are due to shifting validation to the command_validation module - easy fix soon).

File: cylc/flow/task_pool.py [hunk 1]
```diff
@@ -2066,34 +2073,47 @@ def _get_flow_nums(
...
+
+        Arg should have been validated already during command validation.
+
+        Call this method separately for active (n=0) and future tasks.
+        - future tasks: assign the result to the new task
+        - active tasks: merge the result with existing flow numbers
+
+        Note if a single command results in two calls to this method (for
+        active and future tasks), translate --flow=new to an actual flow
+        number first, to avoid incrementing the flow counter twice.
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py:2088]: I don't like this. I've come up with what I think is a more elegant approach at https://github.com/hjoliver/cylc-flow/pull/56

**Review by MetRonnie — CHANGES_REQUESTED** [review]

File: cylc/flow/scripts/trigger.py [hunk 2]
```diff
@@ -17,18 +17,25 @@
...
 
 * Triggering an unqueued waiting task queues it, regardless of prerequisites.
 * Triggering a queued task submits it, regardless of queue limiting.
 * Triggering an active task has no effect (it already triggered).
 
-Incomplete and active-waiting tasks in the n=0 window already belong to a flow.
-Triggering them queues them to run (or rerun) in the same flow.
+Future tasks (n>0) do not already belong to a flow.
+* by default they are assigned to all active flows
+* otherwise, according to the --flow option
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: Technically all `n>0` tasks, for trigger and set purposes, are "future tasks".  If they already ran in the past, that was part of a different, earlier, flow.  Or have I misunderstood you?

File: cylc/flow/task_pool.py [hunk 3]
```diff
@@ -1908,26 +1908,33 @@ def set_prereqs_and_outputs(
...
-            # Existing task proxies.
             self.merge_flows(itask, flow_nums)
             if prereqs:
                 self._set_prereqs_itask(itask, prereqs, flow_nums)
             else:
                 self._set_outputs_itask(itask, outputs)
 
+        # Spawn and set future tasks.
+        flow_nums = self._get_flow_nums(flow, flow_descr, active=False)
         for name, point in future_tasks:
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:1938]: See my earlier comment on "future tasks". Maybe "inactive tasks" is a better way to describe these tasks, but "future tasks" is not incorrect.

File: cylc/flow/task_pool.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/task_pool.py:2088]: I agree it's not ideal. It was kind of a compromise on what I saw as a slightly bigger refactor. But I'll take a look at your way...

File: cylc/flow/scripts/trigger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: OK while I don't think "future" is incorrect, as explained, I agree that "inactive" is better. So I'll change to that. And I've posted a note to Element to check that we all agree.

File: cylc/flow/task_pool.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:1938]: OK while I don't think "future" is incorrect, as explained, I agree that "inactive" is better. So I'll change to that. And I've posted a note to Element to check that we all agree.

File: cylc/flow/scripts/trigger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: (And use of "future_tasks" for non-task-pool tasks, in the code at least, goes back at least 3 years).

File: cylc/flow/scripts/trigger.py [hunk 4]
```diff
@@ -114,6 +122,7 @@ async def run(options: 'Values', workflow_id: str, *tokens_list):
 @cli_function(get_option_parser)
 def main(parser: COP, options: 'Values', *ids: str):
     """CLI for "cylc trigger"."""
+    flow_opts(options.flow, options.flow_wait)
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:125]: Huh, according to me (in the description):
> - remove redundant option checking, now done during command validation

[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:125]: (The trouble with 2-weeks delays, I don't recall my reasoning now, maybe I did forgot about the new command validation regime... I'll fix it).

File: cylc/flow/task_pool.py [hunk 5]
```diff
@@ -1908,26 +1908,33 @@ def set_prereqs_and_outputs(
             flow_descr: description of new flow
 
         """
-        flow_nums = self._get_flow_nums(flow, flow_descr)
-        if flow_nums is None:
-            # Illegal flow command opts
-            return
-
```
Comment (hjoliver) [inline cylc/flow/task_pool.py:1915]: (Maybe this was the "redundant option checking" that I removed, although it seems a bit trivial.)

File: cylc/flow/scripts/trigger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: (Decided on "inactive" anyway)

**Comment by hjoliver** [comment]
(Not done with this yet)

**Comment by oliver-sanders** [comment]
> (Not done with this yet)

Drafting to help with review prioritisation.

File: cylc/flow/scripts/trigger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py]: Done.

**Review by oliver-sanders — APPROVED** [review]

File: cylc/flow/network/schema.py [hunk 6]
```diff
@@ -1998,17 +1998,20 @@ class Arguments:
...
-
-            * {FLOW_ALL} - Triggered tasks belong to all active flows
-              (default).
-            * {FLOW_NEW} - Triggered tasks are assigned to a new flow.
-            * {FLOW_NONE} - Triggered tasks do not belong to any flow.
+            The flow(s) to trigger/set these tasks in.
+
+            By default:
+            * active tasks (n=0) keep their existing flow assignment
+            * inactive tasks (n>0) get assigned all active flows
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:2007]: I guess we need to get our semantics straight. I've taken to saying all inactive tasks, including past ones, are `n>0` (i.e., a positive number of graph edges out from active tasks) but some of them may have run before in this or other flows. Does it particularly help to describe past tasks (in the current flow?) as `n<0`?

File: cylc/flow/scripts/trigger.py [hunk 7]
```diff
@@ -17,18 +17,25 @@
...
-up - will see tasks that ran after triggering event as having run already.
+Active tasks (n=0) already belong to a flow.
+* by default they run in the same flow
+* with --flow=all, they are assigned to all active flows
+* with --flow=INT or --flow=new, the new flow merges with the old one
+* --flow=none is ignored
+
+Note --flow=new increments the global flow counter so if you need multiple
+commands to start a single new flow only use --flow=new in the first command,
+then use the actual new flow number (e.g. read it from the scheduler log).
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:38]: OK I'll go with the indent option. I agree it's somewhat "advanced" but it is also critical if you do start triggering new flows or past tasks.

File: cylc/flow/network/schema.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:2007]: Actually, in line with recent discussions I've changed this to "inactive tasks" which covers past and future.

File: cylc/flow/network/schema.py [hunk 8]
```diff
@@ -1998,17 +1998,20 @@ class Arguments:
...
+
+            Otherwise you can assign (inactive tasks) or add to (active tasks):
+            * a list of integer flow numbers
+            or a single-item list containing one of the following strings:
```
Comment (MetRonnie) [inline cylc/flow/network/schema.py:2011]:
```suggestion
            or one of the following strings:
```

Comment (oliver-sanders) [reply cylc/flow/network/schema.py:2011]: I think it is a list?

Screenshot from 2024-10-14 11-15-30

Comment (MetRonnie) [reply cylc/flow/network/schema.py:2011]: I think "one of the following strings" is clear enough, and won't give the impression you're supposed to type `[all]`

File: tests/integration/test_trigger.py [hunk 9]
```diff
@@ -71,3 +50,199 @@ async def test_trigger_no_flows(one, start, log_filter):
...
+        assert schd.pool._get_task_by_id('1/d').flow_nums == set(
+            [1, 2, 3, 4]
+        )
+        schd.pool.force_trigger_tasks(
+            ['1/e'], flow=[7]
+        )
+        assert schd.pool._get_task_by_id('1/e').flow_nums == set([7])
+
+
+async def test_set(flow, scheduler, start):
```
Comment (MetRonnie) [inline tests/integration/test_trigger.py:158]: I've combined this with the `trigger` test to ensure the behaviours are tested identically for both: https://github.com/hjoliver/cylc-flow/pull/57

**Comment by hjoliver** [comment]
Merged your test tweaks @MetRonnie, tests all passing, should be good to merge now.

**Review by MetRonnie — APPROVED** [review]

**Comment by oliver-sanders** [comment]
@hjoliver

Can you take a look at https://github.com/cylc/cylc-flow/issues/5635 in the light of this PR as it is probably fixed.

**Pull request merged**